# Generación de Texto con modelos GPT

Version Ivan Moran y Erik Vergara

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Ohtar10/icesi-nlp/blob/main/Sesion5/1-text-generation.ipynb)

En este notebook harémos uso de un modelo tipo GPT-2 pre-entrenado en idioma español que utilizaremos para generar texto a partir de un contexto inicial que proveerémos. Luego, harémos fine tuning a este modelo con un dataset de chistes en español y observar como cambia la generación de texto en función del dataset que utilicemos.

#### Referencias
- Dataset: https://huggingface.co/datasets/mrm8488/CHISTES_spanish_jokes
- [Improving Language Understanding by Generative Pre-Training](https://cdn.openai.com/research-covers/language-unsupervised/language_understanding_paper.pdf)
- [Natural Language Processing with Transformers: Building Language Applications With Hugging Face](https://www.amazon.com/Natural-Language-Processing-Transformers-Applications/dp/1098103246)
- [GPT2 Spanish](https://huggingface.co/DeepESP/gpt2-spanish)
- [Fine-Tune a non-Englush GPT-2 Model with Huggingface](https://www.philschmid.de/fine-tune-a-non-english-gpt-2-model-with-huggingface)

# Paquetes y dependencias

Se mantiene la instalacion de paquetes y dependencias.

In [35]:
import pkg_resources
import warnings

warnings.filterwarnings('ignore')

installed_packages = [package.key for package in pkg_resources.working_set]
IN_COLAB = 'google-colab' in installed_packages

In [36]:
#!test '{IN_COLAB}' = 'True' && wget  https://github.com/Ohtar10/icesi-nlp/raw/refs/heads/main/requirements.txt && pip install -r requirements.txt
!test '{IN_COLAB}' = 'True' && sudo apt-get update -y
!test '{IN_COLAB}' = 'True' && sudo apt-get install python3.10 python3.10-distutils python3.10-lib2to3 -y
!test '{IN_COLAB}' = 'True' && sudo update-alternatives --install /usr/local/bin/python python /usr/bin/python3.11 2
!test '{IN_COLAB}' = 'True' && sudo update-alternatives --install /usr/local/bin/python python /usr/bin/python3.10 1
!test '{IN_COLAB}' = 'True' && pip install lightning datasets 'transformers[torch]' sentence-transformers torchinfo evaluate

Hit:1 https://cli.github.com/packages stable InRelease
Hit:2 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease
Hit:3 http://security.ubuntu.com/ubuntu jammy-security InRelease
Hit:4 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease
Hit:5 https://r2u.stat.illinois.edu/ubuntu jammy InRelease
Hit:6 http://archive.ubuntu.com/ubuntu jammy InRelease
Hit:7 http://archive.ubuntu.com/ubuntu jammy-updates InRelease
Hit:8 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease
Hit:9 http://archive.ubuntu.com/ubuntu jammy-backports InRelease
Hit:10 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease
Hit:11 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease
Reading package lists... Done
W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)
Reading packag

# Comparacion de modelos en la generacion de contenido

A continuacion vamos a definir varios modelos para utilizarlos en la generacion de contenido, en este caso, creacion narrativa tipo cuentos, aprovechando el corpus con el que fue entrenado el modelo base DeepESP/gpt2-spanish.

1. DeepESP/gpt2-spanish
2. meta-llama/Llama-3.2-1B
3. API meta-llama/Llama-3.2-1B-Instruct:novita

## 1. Generative pre-training Transformer - GPT

Se mantiene un modelo GPT para poder comparar la generacion de contenido. Sin embargo, es importante tener en cuenta que este modelo (DeepESP/gpt2-spanish) fue entrenado con un corpus de 11,5 GB de textos, que corresponden a 3,5 GB de artículos de Wikipedia y 8 GB de libros (narrativa, cuentos, teatro, poesía, ensayos y divulgación).

### Definicion del modelo

In [37]:
import torch
import transformers
from transformers import AutoModelForCausalLM, AutoTokenizer

device = "cuda" if torch.cuda.is_available() else "cpu"
model_name = "DeepESP/gpt2-spanish"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name).to(device)
model

Loading weights:   0%|          | 0/149 [00:00<?, ?it/s]

The tied weights mapping and config for this model specifies to tie transformer.wte.weight to lm_head.weight, but both are present in the checkpoints, so we will NOT tie them. You should update the config with `tie_word_embeddings=False` to silence this warning
GPT2LMHeadModel LOAD REPORT from: DeepESP/gpt2-spanish
Key                                     | Status     |  | 
----------------------------------------+------------+--+-
transformer.h.{0...11}.attn.masked_bias | UNEXPECTED |  | 
transformer.h.{0...11}.attn.bias        | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


GPT2LMHeadModel(
  (transformer): GPT2Model(
    (wte): Embedding(50257, 768)
    (wpe): Embedding(1024, 768)
    (drop): Dropout(p=0.1, inplace=False)
    (h): ModuleList(
      (0-11): 12 x GPT2Block(
        (ln_1): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (attn): GPT2Attention(
          (c_attn): Conv1D(nf=2304, nx=768)
          (c_proj): Conv1D(nf=768, nx=768)
          (attn_dropout): Dropout(p=0.1, inplace=False)
          (resid_dropout): Dropout(p=0.1, inplace=False)
        )
        (ln_2): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (mlp): GPT2MLP(
          (c_fc): Conv1D(nf=3072, nx=768)
          (c_proj): Conv1D(nf=768, nx=3072)
          (act): NewGELUActivation()
          (dropout): Dropout(p=0.1, inplace=False)
        )
      )
    )
    (ln_f): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
  )
  (lm_head): Linear(in_features=768, out_features=50257, bias=False)
)

### Definicion del Calculo de la Perplejidad

In [38]:
def calcular_perplejidad(model, tokenizer, texto):
    device = next(model.parameters()).device
    inputs = tokenizer(texto, return_tensors="pt").to(device)
    with torch.no_grad():
        outputs = model(**inputs, labels=inputs["input_ids"])
    loss = outputs.loss
    perplexity = torch.exp(loss)
    return perplexity.item()

### Generando texto

Utilizamos el principio del modelo de Completitud estadística, para predecir el siguiente token basándose el corpus literario con el que fue entrenado, para completar una historia, en donde le damos el inicio del cuento.

In [39]:
# 1. Definimos el Prompt con un ejemplo previo para marcar el estilo narrativo
prompt_cuento = """[Estilo: Narrativa fantástica, descriptiva y emocional]

Ejemplo de leyenda breve:
El viejo relojero no reparaba el tiempo, sino los recuerdos. Con un destornillador de plata y una gota de nostalgia, lograba que los péndulos volvieran a marcar el instante exacto en que alguien fue feliz. El taller olía a madera antigua y a momentos recuperados.

Historia corta:
Título: El secreto de la montaña de cristal.
Había una vez en el corazón de un valle olvidado, donde las nubes rozan el suelo y el viento susurra secretos antiguos, un joven buscador de tesoros que"""

inputs = tokenizer(prompt_cuento, return_tensors="pt").to(device)

# 2. Generación optimizada para narrativa
output = model.generate(
    input_ids=inputs["input_ids"],
    attention_mask=inputs["attention_mask"],
    max_new_tokens=100,          # Damos más espacio para que la historia se desarrolle
    do_sample=True,
    temperature=0.9,             # Un poco más bajo que en poesía para mantener la lógica lineal
    top_k=50,
    top_p=0.95,
    repetition_penalty=1.5,      # Evita que repita "Había una vez" o nombres de personajes
    no_repeat_ngram_size=4,      # Evita que repita frases de 3 palabras (clave en cuentos)
    pad_token_id=tokenizer.eos_token_id
)

# 3. Resultado
generated_text_M1 = tokenizer.decode(output[0], skip_special_tokens=True)
print(generated_text_M1)

[Estilo: Narrativa fantástica, descriptiva y emocional]

Ejemplo de leyenda breve:
El viejo relojero no reparaba el tiempo, sino los recuerdos. Con un destornillador de plata y una gota de nostalgia, lograba que los péndulos volvieran a marcar el instante exacto en que alguien fue feliz. El taller olía a madera antigua y a momentos recuperados.

Historia corta:
Título: El secreto de la montaña de cristal.
Había una vez en el corazón de un valle olvidado, donde las nubes rozan el suelo y el viento susurra secretos antiguos, un joven buscador de tesoros que se hace pasar por viajero. 



In [40]:
perplexity_GPT = calcular_perplejidad(model, tokenizer, generated_text_M1)
print(f"Perplexity Model GPT:  {perplexity_GPT}")

Perplexity Model GPT:  65.77442932128906


La perplejidad obtenida para GPT-2 (65.77) sugiere que el modelo presenta una alta incertidumbre al predecir la secuencia generada. Esto indica que el texto producido no se ajusta de manera óptima a la distribución aprendida durante el entrenamiento, lo cual es consistente con la pérdida de coherencia observada en la narrativa.

**Observacion:**

- Se observa que el modelo GPT-2 presenta limitaciones en la coherencia narrativa, evidenciadas en transiciones abruptas y pérdida de continuidad temática. A pesar de haber sido entrenado con un corpus literario amplio, el modelo tiende a priorizar la probabilidad local de los tokens sobre la consistencia global del texto. Esto se refleja en una perplejidad relativamente alta (65.77), lo cual indica una menor capacidad del modelo para ajustar adecuadamente la distribución del lenguaje en el contexto generado.

## 2. Llama 3.2 collection of multilingual large language models (LLMs)

Modelos generativos preentrenados y optimizados para instrucciones en tamaños de 1B y 3B (texto de entrada/texto de salida). Los modelos Llama 3.2, optimizados para texto, están diseñados para casos de uso de diálogo multilingüe, incluyendo tareas de recuperación y resumen de información. Superan a muchos de los modelos de chat de código abierto y propietarios disponibles en los benchmarks habituales del sector.



### Conectarse con Hugging Face

Primero se debe crear una conexion con Hugging Face para descargar el modelo }(Es libre pero se debe autenticar para utilizarlo), sirve tambien para consumir el modelo en fomra de servicio.


In [41]:
import os
from huggingface_hub import login, whoami
import getpass
from google.colab import userdata


# Replace 'YOUR_HF_TOKEN_HERE' with your actual Hugging Face token
os.environ['HF_TOKEN'] = userdata.get('HF_TOKEN')

print("HF_TOKEN environment variable has been set for this session.")

HF_TOKEN environment variable has been set for this session.


In [43]:
token = userdata.get('HF_TOKEN')
login(token)

Note: Environment variable`HF_TOKEN` is set and is the current active token independently from the token you've just configured.


In [44]:
print(whoami())

{'type': 'user', 'id': '6838ab6bdbf608133c533a20', 'name': 'evergarah2000', 'fullname': 'Erik Vergara', 'isPro': False, 'avatarUrl': 'https://cdn-avatars.huggingface.co/v1/production/uploads/no-auth/7if1L4Jb0Ehr-0I0mvjYJ.png', 'orgs': [{'type': 'org', 'id': '6838abec6538ac51b9e6e7c0', 'name': 'ORG-TEST777', 'fullname': 'ORG-TEST', 'avatarUrl': 'https://www.gravatar.com/avatar/35cdfddef36eba98c6d73affedc88b8d?d=retro&size=100'}], 'auth': {'type': 'access_token', 'accessToken': {'displayName': 'NewPractica5', 'role': 'fineGrained', 'createdAt': '2026-03-31T17:17:26.567Z', 'fineGrained': {'canReadGatedRepos': True, 'global': [], 'scoped': [{'entity': {'_id': '66eaebb25fcdca258b9576cb', 'type': 'model', 'name': 'meta-llama/Llama-3.2-1B'}, 'permissions': ['repo.content.read', 'repo.access.read', 'repo.write']}, {'entity': {'_id': '6838ab6bdbf608133c533a20', 'type': 'user', 'name': 'evergarah2000'}, 'permissions': ['repo.content.read', 'repo.access.read', 'repo.write', 'inference.serverless.

### Definicion del modelo

In [45]:
import torch
from transformers import pipeline

# 1. Configuración del modelo Llama 3.2 (1B es ideal para velocidad)
model_id = "meta-llama/Llama-3.2-1B"

pipe = pipeline(
    "text-generation",
    model=model_id,
    torch_dtype=torch.bfloat16,
    device_map="auto"
)

Loading weights:   0%|          | 0/146 [00:00<?, ?it/s]

### Generando texto

In [46]:
# 2. Definición del Prompt Elaborado
# Usamos una estructura que Llama reconoce: Instrucción + Estilo + Inicio
prompt_cuento = """Escribe un cuento corto siguiendo este estilo: Narrativa fantástica, descriptiva y emocional.

Historia:
Título: El secreto de la montaña de cristal.
Había una vez en el corazón de un valle olvidado, donde las nubes rozan el suelo y el viento susurra secretos antiguos, un joven buscador de tesoros que"""

# 3. Generación con parámetros optimizados para Llama
# Nota: Llama 3.2 funciona mejor con una temperatura ligeramente más baja para no perder el hilo
resultados = pipe(
    prompt_cuento,
    max_new_tokens=100,      # Llama puede escribir historias más largas con coherencia
    do_sample=True,
    temperature=0.7,
    top_p=0.9,
    repetition_penalty=1.2,  # Llama es más inteligente, no necesita una penalización tan agresiva como GPT-2
    eos_token_id=pipe.tokenizer.eos_token_id,
    pad_token_id=pipe.tokenizer.eos_token_id
)

# 4. Mostrar el resultado
print(resultados[0]['generated_text'])
generated_text_M2 = resultados[0]['generated_text']


Both `max_new_tokens` (=100) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Escribe un cuento corto siguiendo este estilo: Narrativa fantástica, descriptiva y emocional.

Historia:
Título: El secreto de la montaña de cristal.
Había una vez en el corazón de un valle olvidado, donde las nubes rozan el suelo y el viento susurra secretos antiguos, un joven buscador de tesoros que se encontró con algo más profundo. Era un tesoro sagrado al cual nadie había tenido acceso hasta ese día; era como si lo hubieran robado del propio cielo. Este fue el motivo por el que viajaba a través del bosque mágico llamando a todos los guardianes para que saliesen a buscarlo antes de ser atacados o destruidos. 

A pesar de no poder comunicarse ni hablar entre sí, cada uno


In [47]:
perplexity_Llama = calcular_perplejidad(model, tokenizer, generated_text_M2)
print(f"Perplexity Model Llama:  {perplexity_Llama}")

Perplexity Model Llama:  51.44061279296875


**Observacion:**
- El modelo Llama 3.2 muestra una mejora en la estructuración del texto frente a GPT-2, particularmente en la interpretación del prompt como una instrucción explícita. Sin embargo, aún presenta limitaciones en la continuidad narrativa, evidenciadas en interrupciones del flujo lógico de la historia. Aunque la perplejidad obtenida es menor (51.44), esta métrica no es directamente comparable con GPT-2 debido a diferencias en tokenización y distribución probabilística entre modelos, por lo que la mejora se evalúa principalmente desde un análisis cualitativo.
- La perplejidad calculada para el texto generado por Llama fue evaluada utilizando el modelo GPT-2, lo cual introduce un sesgo metodológico. Esta métrica no refleja la distribución interna del modelo Llama, por lo que no debe interpretarse como una medida directa de su desempeño, sino únicamente como una referencia aproximada bajo el espacio probabilístico de GPT-2.

## 3. API Llama 3.2 de modelos multilingües de lenguaje grande (LLM)

Se realizó una prueba de inferencia remota utilizando el Hugging Face Router API para consumir el modelo Llama-3.2-1B-Instruct. A diferencia de la ejecución local, este enfoque emplea una arquitectura de chat (vía el endpoint de Novita) que permite enviar un prompt estructurado mediante roles de sistema y usuario, optimizando la generación de texto mediante el procesamiento en la nube sin carga directa en el hardware local.

In [48]:
import os
from openai import OpenAI

# 1. Configuración del cliente (Asegúrate de tener HF_TOKEN en tus variables de entorno)
client = OpenAI(
    base_url="https://router.huggingface.co/v1",
    api_key=os.environ["HF_TOKEN"],
)

# 2. Llamada al modelo Llama 3.2 vía API
completion = client.chat.completions.create(
    model="meta-llama/Llama-3.2-1B-Instruct:novita",
    messages=[
        {
            "role": "system",
            "content": "Eres un escritor creativo experto en narrativa fantástica, descriptiva y emocional."
        },
        {
            "role": "user",
            "content": "Escribe un cuento corto titulado 'El secreto de la montaña de cristal'. La historia debe comenzar en el corazón de un valle olvidado donde las nubes rozan el suelo."
        }
    ],
    temperature=0.7,
    max_tokens=500
)

# 3. Impresión del resultado para comparar
print(completion.choices[0].message.content)

**El secreto de la montaña de cristal**

En el corazón de un valle olvidado, donde las nubes rozaban el suelo como lágrimas de un cielo despejado, había una montaña de cristal. La montaña de cristal era un lugar misterioso, un lugar donde la naturaleza parecía haber dejado todo intacto, sin embargo, también parecía haber sido silenciado por la historia.

La montaña de cristal era un lugar de leyendas y misterios. Los habitantes del valle habían oído historias de un lugar donde el cristal era más precioso que el oro, donde los que lo encontraban podrían vivir eternamente. La montaña era considerada un lugar sagrado, un lugar donde la magia se tejió en la atmósfera.

Una noche, una joven llamada Sofía se despertó en su pequeño refugio de campaña, rodeado de montones de lona y tela. La luna estaba a punto de pautar el cielo, y la montaña de cristal se parecía a un monasterio blanco, envuelto en una niebla de incienso. Sofía se sintió atraída por la montaña, como si una partícula de su alm

**Observaciones:**
- El modelo Llama 3.2 en su versión API presenta una mejora significativa en la coherencia y continuidad narrativa. Esto se debe a que utiliza una arquitectura optimizada para instrucciones (Instruct), lo que le permite interpretar correctamente el prompt como una tarea de generación estructurada. A diferencia de los modelos anteriores, mantiene consistencia temática, desarrollo de personajes y progresión lógica en la historia, lo cual evidencia una mejor alineación con tareas de generación controlada.

# Conclusiones

- El modelo GPT-2 evidencia deriva de contexto, al no diferenciar entre el ejemplo proporcionado y la instrucción principal, absorbiendo el contenido semántico en lugar de utilizarlo como guía estilística. Esto refleja una limitación inherente de los modelos autoregresivos no alineados a instrucciones, los cuales no están optimizados para separar contexto, estilo y tarea.
- El modelo Llama 3.2 en ejecución local logra interpretar el inicio del texto como una continuación narrativa, pero presenta limitaciones en la generación de contenido creativo sostenido. Esto se manifiesta en cortes abruptos y falta de desarrollo estructural, lo cual sugiere que, aunque el modelo tiene mejor comprensión del prompt que GPT-2, su capacidad de generación depende fuertemente de la configuración de inferencia y del contexto proporcionado.
- El modelo Llama 3.2 consumido mediante API demuestra un desempeño superior en tareas de generación narrativa, al interpretar correctamente las instrucciones y estructurar el texto de manera coherente. La utilización de un formato de interacción basado en roles (system/user) permite una mejor alineación con la intención del usuario, lo que se traduce en una narrativa más consistente, con inicio, desarrollo y cierre definidos.